app/services/edge_cache/bust.rb

```diff
@@ -1,62 +1,66 @@
 module EdgeCache
   class Bust
-    TIMEFRAMES = [
-      [-> { 1.week.ago }, "week"],
-      [-> { 1.month.ago }, "month"],
-      [-> { 1.year.ago }, "year"],
-      [-> { 5.years.ago }, "infinity"],
-    ].freeze
-
-    def self.call(path)
-      bust(path)
+    def initialize
+      @provider_class = determine_provider_class
     end
 
-    class << self
-      protected
-
-      def bust(path)
-        provider_class = determine_provider_class
-
-        return unless provider_class
+    def self.call(*paths)
+      new.call(*paths)
+    end
 
-        if provider_class.respond_to?(:call)
-          provider_class.call(path)
+    def call(paths)
+      return unless @provider_class
 
-          true
-        else
-          Rails.logger.warn("#{provider_class} cannot be used without a #call implementation!")
-          ForemStatsClient.increment("edgecache_bust.invalid_provider_class",
-                                     tags: ["provider_class:#{provider_class}"])
-          false
-        end
+      paths = Array.wrap(paths)
+      paths.each do |path|
+        @provider_class.call(path)
+        true
+      rescue StandardError => e
+        Honeybadger.notify(e)
+        ForemStatsClient.increment(
+          "edgecache_bust.provider_error",
+          tags: ["provider_class:#{@provider_class}", "error_class:#{e.class}"],
+        )
+        false
```

Comment:
By changing it to an array we need to make a decision on how to interpret `true` and `false` here.

`.each` won't collect the booleans, meaning they aren't effectively used. I'll use an example:

```ruby
[1] pry(main)> paths = %w[a b c]
=> ["a", "b", "c"]
[2] pry(main)> result = paths.each { |p| [true, false].sample }
=> ["a", "b", "c"]
[3] pry(main)> puts result
a
b
c
=> nil
```

The block in `call` is effectively disregarding the return values.

We have a few alternatives, depending if you think we should keep track of them or not for "return value" purposes:

1. use `paths.map()` and collect those and decide what it means to be successful, for example:

```ruby
[4] pry(main)> results = paths.map { |p| [true, false].sample }
=> [true, true, false]
[5] pry(main)> results.all?
=> false
```

In this way we'd return `false` to the caller because at least one cache purge call failed.

2. remove `true` and `false` and ignore the return values (after all we track failures on Datadog)

3. return a report of what happened:

```ruby
[11] pry(main)> results = {}
=> {}
[12] pry(main)> paths.each { |p| results[p] = [true, false].sample }
=> ["a", "b", "c"]
[13] pry(main)> results
=> {"a"=>true, "b"=>true, "c"=>false}
```

What do you think?

Reply:
I'm so glad you brought this up, I meant to bring it up as well. Thank you!

I personally like number 2 - remove the `true` and `false` and ignore the return values. 1) because like you said, we track failures on Datadog and AFAIK, the only times that would/has "failed" is when the service itself is down. And 2) we currently don't have any code to handle the rare case of failure. So even if we return the report (option 3 which is my second favorite), we don't do anything with it. I'd almost rather wait to see what happens. We've gone this long without any code to handle that edge case...I almost would want to wait for it to become an issue so we can understand more as to what's going on. And at that point, we can add code to handle failures and be more informed about what's most useful to return in those cases.

What do you think? I know you mentioned that any choice is fine!

Reply:
Yeah, number 2 is my favorite choice as well. We're not currently doing anything with the return value, if we need to we can take into account that scenario and decide accordingly!